cdc/sink/codec/avro.go

```diff
@@ -193,41 +233,157 @@ func avroEncode(table *model.TableName, manager *AvroSchemaManager, tableVersion
 }
 
 type avroSchemaTop struct {
-	Tp     string                   `json:"type"`
-	Name   string                   `json:"name"`
-	Fields []map[string]interface{} `json:"fields"`
+	Tp        string                   `json:"type"`
+	Name      string                   `json:"name"`
+	Namespace string                   `json:"namespace"`
+	Fields    []map[string]interface{} `json:"fields"`
 }
 
-type logicalType string
+const (
+	tidbType         = "tidb_type"
+	tidbOp           = "_tidb_op"
+	tidbCommitTs     = "_tidb_commit_ts"
+	tidbPhysicalTime = "_tidb_commit_physical_time"
+)
 
-type avroLogicalType struct {
-	Type        string      `json:"type"`
-	LogicalType logicalType `json:"logicalType"`
-	Precision   interface{} `json:"precision,omitempty"`
-	Scale       interface{} `json:"scale,omitempty"`
+var type2TiDBType = map[byte]string{
+	mysql.TypeTiny:       "INT",
+	mysql.TypeShort:      "INT",
+	mysql.TypeInt24:      "INT",
+	mysql.TypeLong:       "INT",
+	mysql.TypeLonglong:   "BIGINT",
+	mysql.TypeFloat:      "FLOAT",
+	mysql.TypeDouble:     "DOUBLE",
+	mysql.TypeBit:        "BIT",
+	mysql.TypeNewDecimal: "DECIMAL",
+	mysql.TypeTinyBlob:   "TEXT",
+	mysql.TypeMediumBlob: "TEXT",
+	mysql.TypeBlob:       "TEXT",
+	mysql.TypeLongBlob:   "TEXT",
+	mysql.TypeVarchar:    "TEXT",
+	mysql.TypeVarString:  "TEXT",
+	mysql.TypeString:     "TEXT",
+	mysql.TypeEnum:       "ENUM",
+	mysql.TypeSet:        "SET",
+	mysql.TypeJSON:       "JSON",
+	mysql.TypeDate:       "DATE",
+	mysql.TypeDatetime:   "DATETIME",
+	mysql.TypeTimestamp:  "TIMESTAMP",
+	mysql.TypeDuration:   "TIME",
+	mysql.TypeYear:       "YEAR",
+}
+
+func getTiDBTypeFromColumn(col *model.Column) string {
+	tt := type2TiDBType[col.Type]
+	if col.Flag.IsUnsigned() && (tt == "INT" || tt == "BIGINT") {
+		return tt + " UNSIGNED"
+	}
+	if col.Flag.IsBinary() && tt == "TEXT" {
+		return "BLOB"
+	}
+	return tt
 }
 
 const (
-	timestampMillis logicalType = "timestamp-millis"
-	timeMillis      logicalType = "time-millis"
-	decimalType     logicalType = "decimal"
+	replacementChar = "_"
+	numberPrefix    = "_"
 )
 
-// ColumnInfoToAvroSchema generates the Avro schema JSON for the corresponding columns
-func ColumnInfoToAvroSchema(name string, columnInfo []*model.Column) (string, error) {
+// sanitizeName escapes not permitted chars for avro
+// debezium-core/src/main/java/io/debezium/schema/FieldNameSelector.java
+// https://avro.apache.org/docs/current/spec.html#names
+func sanitizeName(name string) string {
+	changed := false
+	var sb strings.Builder
+	for i, c := range name {
+		if i == 0 && (c >= '0' && c <= '9') {
+			sb.WriteString(numberPrefix)
+			sb.WriteRune(c)
+			changed = true
+		} else if !(c == '_' ||
+			('a' <= c && c <= 'z') ||
+			('A' <= c && c <= 'Z') ||
+			('0' <= c && c <= '9')) {
+			sb.WriteString(replacementChar)
+			changed = true
+		} else {
+			sb.WriteRune(c)
+		}
+	}
+
+	sanitizedName := sb.String()
+	if changed {
+		log.Warn(
+			fmt.Sprintf(
```

Comment:
this log message can be replaced by `Name is potentially not safe for serialization, replace it, zap.String("name", name), zap.String("replacedName", sanitizedName).